### PR TITLE
Add --list-keys and --delete-key to rpmkeys 

### PR DESCRIPTION
**Pull request by ffesti**
This is a  bit of a hack as it manipulates the parsed cli parameters to
to the "right thing" and then calls rpmcliQuery and rpmErase.

### CODE REVIEW

**Comment by ffesti** [comment]
This feels a bit too quick and dirty. Comments welcome.

File: tools/rpmkeys.c [hunk 1]
```diff
@@ -75,7 +74,29 @@ int main(int argc, char *argv[])
 	break;
     /* XXX TODO: actually implement these... */
     case MODE_DELKEY:
+	struct rpmInstallArguments_s * ia = &rpmIArgs;
+	ARGV_t gpgargs = argvNew();
+	for (char * const * arg = args; *arg; arg++) {
+	    if (strncmp(*arg, "gpg-pubkey-", 11)) {
+		char * gpgarg = rpmExpand("gpg-pubkey-", *arg, NULL);
```
Comment (pmatilai) [inline tools/rpmkeys.c:81]: Why the macro expand? That seems weird/wrong in this context.

File: tools/rpmkeys.c [hunk 2]
```diff
@@ -75,7 +74,29 @@ int main(int argc, char *argv[])
 	break;
     /* XXX TODO: actually implement these... */
```
Comment (pmatilai) [inline tools/rpmkeys.c:75]: Maybe delete this comment while at it? :smile: A fine example of why comments are so bad - they don't get updated along with the code.

**Comment by pmatilai** [comment]
I think I initially wanted to add a sane API for these and then have rpmkeys use that, and when that seemed painful I gave up. But the user wont care *how* it's achieved if it works in a meaningful manner. And, for the user this is far saner than 'rpmkeys --import -> rpm -q -> rpm --erase'. 

Two things:
1) Lets make the option names consistent - either they all should have plural -keys, or none should have it. Since the pre-existing --import doesn't have it, maybe the --delete and --list shouldn't have either? Or, we could add --import-keys as alias to --import.
2) Tests for both added operations.

File: tools/rpmkeys.c [hunk 3]
```diff
@@ -73,9 +72,30 @@ int main(int argc, char *argv[])
...
+	    }
+	}
+	ec = rpmErase(ts, ia, gpgargs);
+	argvFree(gpgargs);
+	break;
     case MODE_LISTKEY:
+	if (args != NULL) {
+	    argerror(_("--list-keys does not take any arguments"));
+	}
+	ARGV_t query = argvSplitString("gpg-pubkey", " ", ARGV_NONE);
```
Comment (pmatilai) [inline tools/rpmkeys.c:94]: It's a bit creative way to initialize an argv... I think
```
ARGV_t query = NULL;
argvAdd(&query, "gpg-pubkey");
```
...would be a well worth the one extra line in obviousness :smile:

**Comment by ffesti** [comment]
Well, I stuck to the names that were in the code...

but it turns out the "keys" in in the name of the utility already and we don't need to repeat that in the cli parameter. It also uses "--delete" while rpm itself uses "--erase". Not sure if that is good because it is a slightly different operation or bad because it is inconsistent. Opinions?

**Comment by pmatilai** [comment]
> Well, I stuck to the names that were in the code...

I know. My not-so-thoroughly thought up names from 10+ years ago, and a fine example of why not to leave such half-assed pieces laying around in the codebase :laughing: 

> It also uses "--delete" while rpm itself uses "--erase". Not sure if that is good because it is a slightly different operation or bad because it is inconsistent. Opinions?

I'm quite sure I chose the names for familiarity from gpg, which has --list-keys and --delete-keys.
I always quite disliked the "erase" terminology in rpm anyhow.

File: tools/rpmkeys.c [hunk 4]
```diff
@@ -42,6 +41,20 @@ static struct poptOption optionsTable[] = {
     POPT_TABLEEND
 };
 
+static ARGV_t gpgkeyargs(ARGV_const_t args) {
+    ARGV_t gpgargs = argvNew();
```
Comment (pmatilai) [inline tools/rpmkeys.c:45]: argvNew() is wholly unnecessary here (it almost always is), just initialize to NULl and argvAdd() and similar will alloc it as needed.

File: tools/rpmkeys.c [hunk 5]
```diff
@@ -42,6 +41,20 @@ static struct poptOption optionsTable[] = {
...
+    for (char * const * arg = args; *arg; arg++) {
+	if (strncmp(*arg, "gpg-pubkey-", 11)) {
+	    char * gpgarg = rpmExpand("gpg-pubkey-", *arg, NULL);
```
Comment (pmatilai) [inline tools/rpmkeys.c:48]: Again, why the macro expand? 'rpmkeys --delete %{somemacro}' seems like a weird use-case to support, I'd think for something like OpenPGP keys you'd rather always be explicit about the key.

Comment (ffesti) [reply tools/rpmkeys.c]: No idea why moving the code elsewhere didn't solve the issue...

**Review by pmatilai — APPROVED** [review]
Looks fine now, thanks.

**Pull request merged**